src/System.Private.ServiceModel/tests/Common/Infrastructure/src/CertificateManager.cs

```diff
@@ -0,0 +1,208 @@
+                    {
+                        try
+                        {
+                            store.Open(OpenFlags.ReadWrite);
+                            store.Add(certificate);
+                            resultCert = certificate;
+                        }
+                        catch (CryptographicException inner)
+                        {
+                            StringBuilder exceptionString = new StringBuilder();
+                            exceptionString.AppendFormat("Error opening StoreName: '{0}' certificate store from StoreLocation '{1}' in ReadWrite mode ", storeName, storeLocation);
+                            exceptionString.AppendFormat("while attempting to install cert with thumbprint '{1}'.", Environment.NewLine, certificate.Thumbprint);
+                            exceptionString.AppendFormat("{0}This is usually due to permissions issues if writing to the LocalMachine location", Environment.NewLine);
+                            exceptionString.AppendFormat("{0}Try running the test with elevated or superuser permissions.", Environment.NewLine);
+
+                            throw new InvalidOperationException(exceptionString.ToString(), inner);
+                        }
+                    }
+                }
+            }
+
+            return resultCert;
+        }
+
+        // Returns the certificate matching the given thumbprint from the given store.
+        // Returns null if not found.
+        private static X509Certificate2 CertificateFromThumbprint(X509Store store, string thumbprint)
+        {
+            X509Certificate2Collection foundCertificates = store.Certificates.Find(X509FindType.FindByThumbprint, thumbprint, validOnly: true);
+            return foundCertificates.Count == 0 ? null : foundCertificates[0];
+        }
+
+        private static X509Certificate2 CertificateFromThumbprint(StoreName storeName,
+                                                                  StoreLocation storeLocation,
+                                                                  string thumbprint)
+        {
+            X509Certificate2 resultCert = null;
+            using (X509Store store = new X509Store(storeName, storeLocation))
+            {
+                store.Open(OpenFlags.ReadOnly);
+                resultCert = CertificateFromThumbprint(store, thumbprint);
+            }
+
+            return resultCert;
+        }
+
+        // Retrieves a root certificate matching the given thumbprint from the root store
+        public static X509Certificate2 RootCertificateFromThumprint(string thumbprint)
+        {
+            return CertificateFromThumbprint(StoreName.Root, PlatformSpecificRootStoreLocation, thumbprint);
+        }
+
+        // Retrieves a root certificate matching the given issuer name and subject name from the root store
+        public static X509Certificate2 RootCertificateFromName(string issuerName, string subjectName)
```

Comment:
The "FromName" methods are new and exist only to support the final check used by the detectors if the install fails.